### ISSUE TITLE
Scala staging not releasing memory

### ISSUE THREAD

**Issue by markehammons** [lampepfl/dotty#12543]
## Compiler version

3.0.0
## Minimized example

```Scala
while true do
  run(iGenAnExpr)
```

## Output

Lots of stuff in the JVM's old gen

## Expectation

Thrown away compiled code is freed.

@nicolasstucki 

I believe the issue is in the QuotesDriver. In the above code, I generate a lot of code via runtime MSP, and then never use it. I had thought this would result in memory being freed, but apparently there's some long-lived storage in the compiler that I haven't found yet, because the above example causes old gen to rapidly fill.

**Comment by markehammons** [comment]
Confirmed in my benchmarking, letting the `Compiler` get garbage collected results in the old generation staying empty. I don't know where, but the staging compiler is hanging on to data much longer than I think it should.

**Comment by smarter** [comment]
Might need to call reset on the contextBase at the end of QuoteDriver#run

**Comment by markehammons** [comment]
@smarter yes, this is the solution

I'll create an MR for it

**Comment by smarter** [comment]
Get a heap dump with visualvm to see what it's holding to?

**Comment by markehammons** [comment]
It seems that there's a couple of SymDenotations$ClassDenotation instances holding on to a huge portion of memory. One of these is 322MB.

**Comment by smarter** [comment]
can you upload a heap dump?

**Comment by markehammons** [comment]
Yes I’ll upload one tomorrow. Where should I put it, and does it matter if it’s like 4GB?

**Comment by smarter** [comment]
That's fine if you find a place to upload it :). But a smaller heap that reproduces the issue would be nice too.

**Comment by markehammons** [comment]
I've uploaded the file to google drive. It's not the biggest heap dump, but it should suffice.

https://drive.google.com/file/d/1qRguIJ6WcJag-M4klEendbsLQez6G0Ri/view?usp=sharing

**Comment by nicolasstucki** [comment]
@markehammons What is the code you are generating in `iGenAnExpr`?

**Comment by nicolasstucki** [comment]
What is the full code? When do you expect to release the memory?

**Comment by nicolasstucki** [comment]
Each compiler instance caches the BaseContext [here](https://github.com/lampepfl/dotty/blob/master/staging/src/scala/quoted/staging/QuoteDriver.scala#L24). This is necessary if we do not want to reload every definition from the class files each time we call `run`. Maybe that is the memory that you see. In theory, it should be freed if the compiler instance is dropped.

**Comment by smarter** [comment]
> Each compiler instance caches the BaseContext here.

Yes, that's why I suggested calling reset on this context earlier in the discussion, but that doesn't seem to be enough. Looking at the heap dump it seems that all the SymDenotations which are retained are TermRef called "Vector_this" (I assume generated by inlining) which show up in the baseTypeCache of the Vector class. Maybe the issue is that baseTypeCache is caching transient terms generated by inlining which don't survive afterwards? But yeah we'd have to see the code to know.

**Comment by markehammons** [comment]
The source code is as follows:

JMH Benchmark:
```scala
  @Benchmark @OutputTimeUnit(TimeUnit.SECONDS)
  def compileSpeed =
    run(sumGen[Int](mVec21.size))
```
# Sumgen and friends
```scala
def fillArray[T](seq: Seq[Expr[T]])(using Quotes)(using Type[T]): Expr[Array[T]] = seq match
  case Seq('{ $t: Int}, tail @ _*) => 
    val arrAssign = for
      ('{$exp: Int},index) <- seq.zipWithIndex
    yield (arr: Expr[Array[Int]]) => '{
      $arr(${Expr(index)}) = $exp
    }
    '{
      val array = new Array[Int](${Expr(seq.size)})
      ${Expr.block(arrAssign.map(_('{array})).toList, '{array})}
    }.asExprOf[Array[T]]

def genSum2[T](
  arity: Int,
  v1: Expr[Vector[T]],
  v2: Expr[Vector[T]],
  ord: Expr[Ordering[T]],
  num: Expr[Numeric[T]],
  ct: Expr[ClassTag[T]]
)(using Type[T])(using Quotes): Expr[Vector[T]] =

  val addExpr: (Expr[T], Expr[T]) => Expr[T] = {
    case ('{ $x: Int }, '{ $y: Int }) => '{($x + $y)}.asExprOf[T]
    case (x,y) => '{$num.plus($x, $y)}
  }

  val exprs = for
    n <- 0 until arity
    nExpr = Expr(n)
  yield addExpr('{$v1($nExpr)}, '{$v2($nExpr)})
  
  val arityExpr = Expr(arity)

  val arrayExpr: Expr[Seq[T]] => Expr[Array[T]] = {
    case '{ $s: Seq[Int] } => '{Array.apply(${s.asExprOf[Seq[Int]]}*)}.asExprOf[Array[T]]
    case '{ $s: Seq[T] } => '{Array[T]($s*)(using $ct)}
  }

  '{
    new Vector(ArraySeq.unsafeWrapArray(${fillArray(exprs)}))(using $num)
  }

def sumGen[T](arity: Int)(using Quotes)(using Type[T]): Expr[(Ordering[T], Numeric[T], ClassTag[T]) ?=> (Vector[T], Vector[T]) => Vector[T]] = '{
  (ord: Ordering[T], num: Numeric[T], ct: ClassTag[T]) ?=> (v1: Vector[T], v2: Vector[T]) =>
   ${genSum2[T](arity, '{v1}, '{v2}, '{ord}, '{num}, '{ct})}
}
```

**Comment by nicolasstucki** [comment]
Calling `reset` on the base context is not enough, the leak comes from [`val definitions`](https://github.com/lampepfl/dotty/blob/master/compiler/src/dotty/tools/dotc/core/Contexts.scala#L848). If we reset it we avoid the memory problems but performance is affected.

The `X` axis shows each call to run and the `Y` the ms it took to run. It is 3X faster with the cache.

<img width="733" alt="Screenshot 2021-06-03 at 15 37 48" src="https://user-images.githubusercontent.com/3648029/120654881-64bf8e00-c482-11eb-836d-4603e123b96c.png">

**Comment by nicolasstucki** [comment]
## Scenario 1: Cached `Compiler`
```scala
given toolbox: Compiler = Compiler.make(this.getClass.getClassLoader)
while true do
  run { ... }
```
* 3X faster to execute each `run`.
* More GC CPU activity
* Compiler cache can be large

<img width="682" alt="Screenshot 2021-06-03 at 15 56 59" src="https://user-images.githubusercontent.com/3648029/120657302-a8b39280-c484-11eb-9576-17023db51657.png">
<img width="691" alt="Screenshot 2021-06-03 at 15 54 13" src="https://user-images.githubusercontent.com/3648029/120657337-b2d59100-c484-11eb-825b-83355efcab76.png">



## Scenario 2: Fresh `Compiler`
```scala
while true do
  given toolbox: Compiler = Compiler.make(this.getClass.getClassLoader)
  run { ... }
```
* 3X slower to execute each `run`.
* Almost no GC CPU activity
* Low memory footprint

<img width="697" alt="Screenshot 2021-06-03 at 15 57 40" src="https://user-images.githubusercontent.com/3648029/120657374-bb2dcc00-c484-11eb-92e5-1c60227206f7.png">
<img width="690" alt="Screenshot 2021-06-03 at 15 57 30" src="https://user-images.githubusercontent.com/3648029/120657387-bcf78f80-c484-11eb-9230-f3ba7cd7c9de.png">

**Comment by nicolasstucki** [comment]
Not sure if we should change anything in the caching mechanism for staging. It provides the flexibility to either cache and run fast, or not cache and run a bit slower without consuming to much memory.

The optimizations we should be looking at are in the Context itself which are things that also affect the normal compiler. I saw may trees retained which are probably trees attached to symbols, these could in theory be dropped and recomputed.

**Comment by smarter** [comment]
> Calling reset on the base context is not enough, the leak comes from val definitions.

More precisely it comes from the baseTypeCache of some denotations: https://github.com/lampepfl/dotty/issues/12543#issuecomment-849607084 which could be turned into a WeahHashMap (and also maybe shouldn't cache termrefs at all?)
